### ISSUE TITLE
non_exhaustive_omitted_patterns should not warn matches! macro

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust#117304]
I tried this code:

```rust
#![feature(non_exhaustive_omitted_patterns_lint)]
#![warn(non_exhaustive_omitted_patterns)]

use std::sync::atomic::Ordering; // #[non_exhaustive] enum

pub fn is_seqcst(order: Ordering) -> bool {
    matches!(order, Ordering::SeqCst)
}
```

I expected to see this happen: no warning

Instead, this happened: 

```
warning: some variants are not matched explicitly
 --> src/lib.rs:7:14
  |
7 |     matches!(order, Ordering::SeqCst)
  |              ^^^^^ patterns `std::sync::atomic::Ordering::Relaxed`, `std::sync::atomic::Ordering::Release`, `std::sync::atomic::Ordering::Acquire` and 1 more not covered
  |
  = help: ensure that all variants are matched explicitly by adding the suggested match arms
  = note: the matched value is of type `std::sync::atomic::Ordering` and the `non_exhaustive_omitted_patterns` attribute was found
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1b2d1a14933830272c70b8736f7a7a0a)

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (2f1bd0729 2023-10-27)
binary: rustc
commit-hash: 2f1bd0729b74787f55d4cbc7818cfd787cd43a99
commit-date: 2023-10-27
host: aarch64-apple-darwin
release: 1.75.0-nightly
LLVM version: 17.0.3
```

---

@rustbot label +F-non_exhaustive_omitted_patterns_lint

**Comment by clarfonthey** [comment]
This feels like something that could be easily solved by just adding an `allow` for the lint inside the macro. I can't think of a situation where you would want this lint to fire here.

**Comment by Jules-Bertholet** [comment]
I attempted to add the `allow`, but it turns out to not be so simple. Log of my struggles:

1. Add `#[allow( allow(non_exhaustive_omitted_patterns)]` above the `match`, inside the macro: compiler complains that `non_exhaustive_omitted_patterns` is unstable
2. Add `#[allow_internal_unstable(non_exhaustive_omitted_patterns_lint)]`: doesn't change anything
3. Patched compiler to make `allow_internal_unstable` work with lints: doesn't help the bootstrap compiler
4. Use  `#[cfg_attr(not(bootstrap), allow(non_exhaustive_omitted_patterns)]` instead: compiler complains that `stmt_expr_attributes` is unstable
5. Add `#[allow_internal_unstable(stmt_expr_attributes)]`: but `stmt_expr_attributes` doesn't work with `allow_internal_unstable` either
6. Patch compiler again to make `stmt_expr_attributes` work with `allow_internal_unstable`: but that does nothing for the bootstrap compiler
7. A `cfg_attr`ed out attribute still requires `stmt_expr_attributes`, so `#[cfg_attr(not(bootstrap), ...]` can't save us here. (I gave up at this point)

@rustbot label -E-easy

**Comment by Nadrieril** [comment]
Sounds like stabilizing the lint would solve this issue. It's a weird lint though, I wouldn't want to stabilize it as-is tho, I'd prefer something more principled like [this](https://github.com/rust-lang/rust/issues/89554#issuecomment-1846909652).

**Comment by Jules-Bertholet** [comment]
6 and 7 from my original comment were resolved by #117420, so someone could presumably pick this up again…

**Comment by jplatte** [comment]
> Patched compiler to make `allow_internal_unstable` work with lints

You didn't PR this part though, right? Do you still have the patch somewhere?

**Comment by Jules-Bertholet** [comment]
> Do you still have the patch somewhere?

https://github.com/Jules-Bertholet/rust/commit/257692d8bd0c1da709c37c87d85aaa9ffec39f19
